### PR TITLE
Shatter glass by slapping the floor

### PR DESCRIPTION
**Pull request by nightred**
## About The Pull Request

Glass shards are usable in many applications but getting them is a chore, now you can slap the floor with a glass sheet and it will shatter.
This means no more making a table and jumping on it to make a glass shard.
Also added in hand crafting recipe if that is your thing.

This will only break one sheet at a time.

## Why It's Good For The Game

Glass shard are useful, simpler access to them helps.

## Changelog
:cl:
add: Slapping the floor with glass will shatter the glass.
/:cl:

### CODE REVIEW

**Comment by ATH1909** [comment]
Thank you for PRing this suggestion, nightred, it means a lot.

File: code/game/objects/items/stacks/sheets/sheets.dm [hunk 1]
```diff
@@ -14,8 +14,27 @@
...
+ * The inital use case is glass sheets breaking in to shards when the floor is hit.
+ * Args:
+ * * user: The user that did the action
+ * * params: paramas passed in from attackby
+ */
+/obj/item/stack/sheet/proc/attacked_by_floor(mob/user, params)
+	if(!shards_to)
+		return
+	use(1)
+	to_chat(user, "<span class='notice'>You shatter one [name].</span>")
```
Comment (ATH1909) [inline code/game/objects/items/stacks/sheets/sheets.dm:38]:
```suggestion
	to_chat(user, "<span class='notice'>You shatter one sheet of [name].</span>")
```

**Comment by nemvar** [comment]
How about you just add the shards to the crafting options? This seems really clunky. I see people accidentally breaking sheets when trying to click machine frames.

**Comment by nightred** [comment]
> How about you just add the shards to the crafting options? This seems really clunky. I see people accidentally breaking sheets when trying to click machine frames.

I like the idea of in world crafting more then menus, slap crafting feels more fun.

File: code/game/turfs/open/floor.dm [hunk 2]
```diff
@@ -161,13 +161,16 @@
...
+/turf/open/floor/attackby(obj/item/object, mob/user, params)
+	if(!object || !user)
 		return 1
 	if(..())
 		return 1
-	if(intact && istype(C, /obj/item/stack/tile))
-		try_replace_tile(C, user, params)
+	if(intact && istype(object, /obj/item/stack/tile))
+		try_replace_tile(object, user, params)
+	if(istype(object, /obj/item/stack/sheet))
```
Comment (Rohesie) [inline code/game/turfs/open/floor.dm:171]:
```suggestion
	if(intact && istype(object, /obj/item/stack/tile))
		try_replace_tile(object, user, params)
		return FALSE
	if(istype(object, /obj/item/stack/sheet))
```

Comment (nightred) [reply code/game/turfs/open/floor.dm]: Fixed this up

File: code/game/objects/items/stacks/sheets/sheets.dm (hunk 1 above)
Comment (nightred) [reply code/game/objects/items/stacks/sheets/sheets.dm]: Added the wording

File: code/game/turfs/open/floor.dm [hunk 3]
```diff
@@ -163,15 +163,17 @@
 
 /turf/open/floor/attackby(obj/item/object, mob/user, params)
 	if(!object || !user)
-		return 1
+		return TRUE
 	if(..())
-		return 1
+		return TRUE
```
Comment (Rohesie) [inline code/game/turfs/open/floor.dm]:
```cpp
. = ..()
if(.)
	return
```

File: code/game/turfs/open/floor.dm [hunk 4]
```diff
@@ -161,14 +161,19 @@
...
-		return 1
+/turf/open/floor/attackby(obj/item/object, mob/user, params)
+	if(!object || !user)
+		return TRUE
 	if(..())
-		return 1
-	if(intact && istype(C, /obj/item/stack/tile))
-		try_replace_tile(C, user, params)
-	return 0
+		return TRUE
```
Comment (Rohesie) [inline code/game/turfs/open/floor.dm:168]:
```cpp
. = ..()
if(.)
	return
```

File: code/game/turfs/open/floor.dm [hunk 5]
```diff
@@ -161,14 +161,19 @@
...
+	if(intact && istype(object, /obj/item/stack/tile))
+		try_replace_tile(object, user, params)
+		return FALSE
+	if(istype(object, /obj/item/stack/sheet))
+		var/obj/item/stack/sheet/sheets = object
+		sheets.attacked_by_floor(user, params)
+		return FALSE
```
Comment (Rohesie) [inline code/game/turfs/open/floor.dm:175]: Personally I think that returning the `attacked_by_floor` return is the best
That said, are you sure you want to return `FALSE` when the attack chain should be successful here?

File: code/game/objects/items/stacks/sheets/sheets.dm [hunk 6]
```diff
@@ -14,8 +14,27 @@
...
+/**
+ * Facilitates sheets being smacked on the floor
+ *
+ * This is used for crafting by hitting the floor with items.
+ * The inital use case is glass sheets breaking in to shards when the floor is hit.
+ * Args:
+ * * user: The user that did the action
+ * * params: paramas passed in from attackby
+ */
+/obj/item/stack/sheet/proc/attacked_by_floor(mob/user, params)
```
Comment (Rohesie) [inline code/game/objects/items/stacks/sheets/sheets.dm:34]: Are you sure `attacked_by_floor` is the most clear name for something called while attacking the floor?

File: code/game/turfs/open/floor.dm (hunk 3 above)
Comment (nightred) [reply code/game/turfs/open/floor.dm]: you are right

File: code/game/objects/items/stacks/sheets/sheets.dm (hunk 6 above)
Comment (nightred) [reply code/game/objects/items/stacks/sheets/sheets.dm:34]: Finding a better name now.

File: code/game/turfs/open/floor.dm (hunk 4 above)
Comment (nightred) [reply code/game/turfs/open/floor.dm:168]: yes will happen

File: code/game/turfs/open/floor.dm (hunk 5 above)
Comment (nightred) [reply code/game/turfs/open/floor.dm:175]: I see what you are saying, i will rework this.

**Comment by nightred** [comment]
I added in hand crafting recipes because options are good.

File: code/game/turfs/open/floor.dm [hunk 7]
```diff
@@ -161,14 +161,19 @@
...
-		return 1
-	if(intact && istype(C, /obj/item/stack/tile))
-		try_replace_tile(C, user, params)
-	return 0
+/turf/open/floor/attackby(obj/item/object, mob/user, params)
+	if(!object || !user)
+		return TRUE
+	. = ..()
+	if(.)
+		return TRUE
```
Comment (Rohesie) [inline code/game/turfs/open/floor.dm:169]:
```suggestion
		return
```
If this is truthy, let it pass whatever return the parent sent.

**Comment by Rohesie** [comment]
Looks good to me, after that small change request.
How about for you, @nemvar ?

**Comment by nemvar** [comment]
I don't like it because I see myself breaking glass when I just try to click glass on a machine frame. Should at least be limited to walls or harm intent.

File: code/game/turfs/open/floor.dm (hunk 7 above)
Comment (nightred) [reply code/game/turfs/open/floor.dm:169]: yes i have no idea why i did not think of that.

**Comment by nightred** [comment]
let me make that change nemvar, i like it.

**Comment by Qustinnus** [comment]
the shards_to var belongs in mat datums

**Comment by Wallemations** [comment]
I personally enjoy the idea, although I still might find myself accidentally breaking the glass. I would have suggested making it that you need to attack the glass with a toolbox or something to break it, however your version still offers a better alternative than placing a halftile window and breaking that.

**Comment by trerri** [comment]
What if this happened when you throw a glass sheet? Imagine throwing a window panel in real life. Of course wouldn't happen if it's reinforced

**Pull request merged**